### PR TITLE
Julian calendar

### LINKED ISSUE #1119: Implement the Coptic calendar
Part of https://github.com/unicode-org/icu4x/issues/534

This will require implementing a Julian calendar to wrap around

### LINKED ISSUE #1350: Fix iso offset for leap years
```
fn test_offset_with_leap_years() {
    let today = Date::new_iso_date_from_integers(2020, 2, 28).unwrap();
    let today_plus_2 = Date::new_iso_date_from_integers(2020, 3, 1).unwrap();
    let offset = today.clone().added(DateDuration::new(0, 0, 0, 2));
    assert_eq!(offset, today_plus_2);
}
```

test fails with the following output
```
left: `Date(IsoDateInner { day: IsoDay(30), month: IsoMonth(2), year: IsoYear(2020) }, for calendar ISO)`,
right: `Date(IsoDateInner { day: IsoDay(1), month: IsoMonth(3), year: IsoYear(2020) }, for calendar ISO)`',
```

### PR DESCRIPTION
**Pull request by pandusonu2**
Partially fixes #1119

### CODE REVIEW

File: components/calendar/src/julian.rs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+pub struct Julian;
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+// The inner date type used for representing Date<Georgian>
+pub struct JulianDateInner(IsoDateInner);
+
+impl Calendar for Julian {
+    type DateInner = JulianDateInner;
+    fn date_from_iso(&self, iso: Date<Iso>) -> JulianDateInner {
+        *iso.inner()
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:18]: issue: this and the next function will need to convert: you're using IsoDateInner for convenience here but it's worth noting that the inner value is not actually an ISO compatible date.

File: components/calendar/src/julian.rs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    fn days_in_month(&self, date: &Self::DateInner) -> u8 {
+        Self::days_in_month(date.year, date.month);
+    }
+
+    fn day_of_week(&self, date: &Self::DateInner) -> types::IsoWeekday {
+        let leap_years_offset = date.year.0 / 4;
+        let days_to_current_year = 365 * date.year.0 + leap_years;
+        let year_offset = days_to_current_year % 7;
+
+        let month_offset = if Self::is_leap_year(date.year) {
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:46]: Given that the _only_ thing different between these functions is `is_leap_year`, i wonder if some of this code can be shared on ISODateInner and have it be parametrized over `F: Fn(Year) -> bool` for the `is_leap_year` bit.

Comment (Manishearth) [reply components/calendar/src/julian.rs:46]: This would make it trivial to reimplement offset as well

Comment (Manishearth) [reply components/calendar/src/julian.rs:46]: It's actually probably better to generalize the offset function, and for weekday just convert to ISO first. This could probably be a default implementation on the `Calendar` trait, even.

Comment (pandusonu2) [reply components/calendar/src/julian.rs:46]: Does making leap_year as a trait that has to be implemented for Calendar make sense?

Comment (Manishearth) [reply components/calendar/src/julian.rs:46]: No, because other calendars do this differently. I think for now we should make the _functions_ generic, and think about unifying further later.

For example, the Indian National (Shaka) calendar probably can share some code as well: It's a calendar _like_ the Gregorian/Julian ones but has different month lengths.

However the Islamic and Hebrew calendars work totally differently and cannot share much code. Perhaps the offset function can be shared but it will not be as easy.

Comment (Manishearth) [reply components/calendar/src/julian.rs:46]: My general hope here is that we incrementally make things more generic: So here we should make the offset/until functions sufficiently generic to support ISO and Julian (but not other calendars, and Gregorian/Buddhist can rely on ISO), but not further. Then when we add the Indian National Calendar we should figure out how to make `offset()` and `until()` more generic.

Comment (pandusonu2) [reply components/calendar/src/julian.rs:46]: Ok, for now, just copying it from iso and reimplementing it for julian

Comment (Manishearth) [reply components/calendar/src/julian.rs:46]: Sounds good.

File: components/calendar/src/julian.rs [hunk 3]
```diff
@@ -74,17 +74,50 @@ impl Calendar for Julian {
...
+    #[allow(clippy::field_reassign_with_default)]
     fn until(
         &self,
         date1: &Self::DateInner,
         date2: &Self::DateInner,
         largest_unit: DateDurationUnit,
         smallest_unit: DateDurationUnit,
     ) -> DateDuration<Self> {
-        // TODO
+        Iso.until(&date1.0, &date2.0, largest_unit, smallest_unit).cast_unit()
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:120]: This isn't correct since the difference between ISO Feb 1 and ISO March 1 on a year which is leap in Julian but not Gregorian (e.g. 1900) is 29 days in Julian and 28 in Gregorian

File: components/calendar/src/julian.rs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+//! This module contains types and implementations for the Julian calendar
+
+#[derive(Copy, Clone, Debug, Default)]
+// The Julian calendar
+pub struct Julian;
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+// The inner date type used for representing Date<Georgian>
+pub struct JulianDateInner(IsoDateInner);
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:13]: I wonder if we should have an independent Inner type so that it's not easy to accidentally mix ISO and Julian values. We would need JulianDay/Month/Year types as well. The purpose of the type safety was to make mistakes where algorithms are mixed impossible.

I don't think we should do this _now_, but it's a thing we can do after the offset/until functions have been abstracted away.

I suspect we can have internal `Day<Calendar>`, `Month<Calendar>`, etc types and a shared `BasicDateInner<Calendar>` type that has the trait magic.

File: components/calendar/src/julian.rs (hunk 3 above)
Comment (pandusonu2) [reply components/calendar/src/julian.rs:120]: I think the current impl would return the difference as 1 month for ISO, rather than 28/29 days 🤔  

On checking the iso, I only see that the year, month and day are subtracted and that value is returned in an object. For that type of implementation, I think it is kinda ok to reuse it.

But yeah, its kinda wrong impl, will do julians own impl

Comment (Manishearth) [reply components/calendar/src/julian.rs:120]: Both the offset and until functions should be generically extracted so they can share an implementation

Comment (pandusonu2) [reply components/calendar/src/julian.rs:120]: Sure, so keeping this as it is then.

File: components/calendar/src/iso.rs [hunk 5]
```diff
@@ -14,10 +14,10 @@ pub struct Iso;
 
 /// A 1-indexed representation of an ISO day
 #[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
-pub struct IsoDay(u8);
+pub struct IsoDay(pub u8);
```
Comment (Manishearth) [inline components/calendar/src/iso.rs:17]: issue: These shouldn't be made public: they're private to maintain the invariants on the inner field. Feel free to add From impls to integers if needed.

File: components/calendar/src/iso.rs [hunk 6]
```diff
@@ -183,8 +183,8 @@ impl Calendar for Iso {
             }
         };
 
-        let january_1_2000 = 5; // Saturday
-        let day_offset = (january_1_2000 + year_offset + month_offset + date.day.0 as i32) % 7;
+        let march_1_200 = 5; // Saturday
```
Comment (Manishearth) [inline components/calendar/src/iso.rs:186]: suggestion: I don't think `day_of_week` needs to be changed: For other calendars we should just use a default implementation of converting to Iso and calling `day_of_week` on it. For Julian I think the following things need to be different from gregorian

 - code for conversion to/from iso
 - code for handling offsets and subtraction. This should be shared between calendars as already mentioned

File: components/calendar/src/julian.rs [hunk 7]
```diff
@@ -179,6 +132,46 @@ impl Julian {
 
         prev_month_days + date.day.0 as u32
     }
+
+    fn calculate_slack_from_iso(date: IsoDateInner) -> u32 {
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:136]: What's this function attempting to do?

**Review by Manishearth — CHANGES_REQUESTED** [review]

File: components/calendar/src/iso.rs (hunk 6 above)
Comment (pandusonu2) [reply components/calendar/src/iso.rs:186]: This doesn't really change the implementation right?! It just changes the date used to measure, how will that effect anything elsewhere? 🤔 

Julian and Georgian are on the same day this day, so it kinda makes things convenient

File: components/calendar/src/julian.rs (hunk 7 above)
Comment (pandusonu2) [reply components/calendar/src/julian.rs:136]: To calculate what is the difference between georgian and julian calendars.

March 1st 200 should give -> 0
today() should give -> 13

File: components/calendar/src/iso.rs (hunk 6 above)
Comment (Manishearth) [reply components/calendar/src/iso.rs:186]: Well you have to introduce the whole slack-calculation step as well.

It's far more straightfoward to keep this impl as is and have Julian (and all the other calendars) just convert to ISO before calling this

File: components/calendar/src/julian.rs (hunk 7 above)
Comment (Manishearth) [reply components/calendar/src/julian.rs:136]: Gotcha. I suspect this can simply be the implementation code in the to/from iso conversion functions in the `Calendar` impl?

**Comment by pandusonu2** [comment]
@Manishearth sorry for force push, just redid the PR, so would be better to check from start.

TODO: Add test cases

File: components/calendar/src/julian.rs [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+use crate::{types, Calendar, Date, DateDuration, DateDurationUnit, DateTime, DateTimeError};
+use core::convert::TryInto;
+
+#[derive(Copy, Clone, Debug, Default)]
+// The Julian calendar
+pub struct Julian;
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+// The inner date type used for representing Date<Georgian>
+pub struct JulianDateInner(IsoDateInner);
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:17]: issue: please document whether the inner date numbers belong to the julian date or its iso-converted version

File: components/calendar/src/julian.rs [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+    fn days_in_month(year: IsoYear, month: IsoMonth) -> u8 {
+        match u8::from(month) {
+            4 | 6 | 9 | 11 => 30,
+            2 if Self::is_leap_year(year) => 29,
+            2 => 28,
+            _ => 31,
+        }
+    }
+
+    fn calculate_day_difference_between_calendars(date: IsoDateInner) -> i32 {
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:154]: issue: please document what the input and output should be

File: components/calendar/src/julian.rs [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+    fn date_from_iso(&self, iso: Date<Iso>) -> JulianDateInner {
+        let added_days = Self::calculate_day_difference_between_calendars(*iso.inner());
+        Self::offset_date(
+            &mut JulianDateInner(*iso.inner()),
+            DateDuration::new(0, 0, 0, -added_days),
+        )
+    }
+
+    fn date_to_iso(&self, date: &Self::DateInner) -> Date<Iso> {
+        let added_days = Self::calculate_day_difference_between_calendars(date.0);
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:30]: question: I'm unsure how this manages to work _both_ ways, since this function seems to take an ISO date and tell you what to add to it to get the Julian date, the reverse should not automatically work? I guess that's what the edge case is dealing with, but all of this is kinda unclear, can you link to sources or perhaps factor this as two separate functions?

File: components/calendar/src/julian.rs [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+
+    #[allow(clippy::field_reassign_with_default)]
+    fn until(
+        &self,
+        date1: &Self::DateInner,
+        date2: &Self::DateInner,
+        largest_unit: DateDurationUnit,
+        smallest_unit: DateDurationUnit,
+    ) -> DateDuration<Self> {
+        Iso.until(&date1.0, &date2.0, largest_unit, smallest_unit)
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:81]: issue: we can't actually use this because the durations mean different things

we should do the same code copying as offset_date here

File: components/calendar/src/julian.rs [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+        let slack = year / 100 - year / 400 - 2;
+
+        if year % 100 == 0 && year % 400 != 0 && u8::from(date.month) <= 2 {
+            slack - 1
+        } else {
+            slack
+        }
+    }
+
+    fn offset_date(date: &mut JulianDateInner, mut offset: DateDuration<Self>) -> JulianDateInner {
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:166]: thought: I'd really like this code to be refactored out and made common, but perhaps I can do this for you once the code lands? This will be useful for the indian national calendar as well

**Review by Manishearth — CHANGES_REQUESTED** [review]
I think this is good! I would prefer if the offsetting code was shared, and there's some confusion around exactly what the logical types are, but we should be able to land this soon!

File: components/calendar/src/julian.rs [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+use crate::iso::{Iso, IsoDateInner, IsoDay, IsoMonth, IsoYear};
+use crate::{types, Calendar, Date, DateDuration, DateDurationUnit, DateTime, DateTimeError};
+use core::convert::TryInto;
+
+#[derive(Copy, Clone, Debug, Default)]
+// The Julian calendar
+pub struct Julian;
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+// The inner date type used for representing Date<Georgian>
```
[the remark this answers is not in the document]
Comment (pandusonu2) [reply components/calendar/src/julian.rs:16]: Done

File: components/calendar/src/julian.rs (hunk 8 above)
Comment (pandusonu2) [reply components/calendar/src/julian.rs:17]: Did comment, not sure if that is clear enough

File: components/calendar/src/julian.rs (hunk 9 above)
Comment (pandusonu2) [reply components/calendar/src/julian.rs:154]: Done

File: components/calendar/src/julian.rs (hunk 12 above)
Comment (pandusonu2) [reply components/calendar/src/julian.rs:166]: Maybe I can pick up the refactoring out as well if we know for sure where the helper methods are supposed to be.

File: components/calendar/src/julian.rs (hunk 10 above)
Comment (pandusonu2) [reply components/calendar/src/julian.rs:30]: I dont really have any documentation on it, wrote the code based on my calculations sort of, so this might as well be wrong.

> perhaps factor this as two separate functions?

Which part specifically?! I think separating out edge case to a separate method might make it harder to read, as the edge case is part of the core logic

Comment (Manishearth) [reply components/calendar/src/julian.rs:31]: I guess it's fine if it's in a single function, but please add comments explaining why this is needed

Or ideally look up an algorithm for each direction and implement that.

Comment (pandusonu2) [reply components/calendar/src/julian.rs:36]: Added comments

Comment (Manishearth) [reply components/calendar/src/julian.rs:36]: So I think the thing that concerns me about this is the use of `calculate_day_difference()`: what if it is called around the boundary when the year is different in ISO and Julian and each year has a different day difference.

I would really be more comfortable if there were some sources cited on this.

Comment (Manishearth) [reply components/calendar/src/julian.rs:36]: @sffc thoughts?

@pandusonu2 if you need, code from the calendar arithmetic book has been published and the relevant code is https://github.com/EdReingold/calendar-code2/blob/1ee51ecfaae6f856b0d7de3e36e9042100b4f424/calendar.l#L1689-L1738 , though it's in Lisp. It's okay to refer to this code

Comment (pandusonu2) [reply components/calendar/src/julian.rs:36]: I added test cases for the same, if that helps

Also one of the source I found is https://smallbusiness.chron.com/convert-julian-iso-49137.html perhaps?! but it doesnt handle the edge case.

The above (N3/4 - 2) is equivalent to (n/100 - n/400 - 2) which is being done in the calculate_day_difference()

Comment (pandusonu2) [reply components/calendar/src/julian.rs:36]: Also, the main resource I used for this conversion is actually: https://www.fourmilab.ch/documents/calendar/ 
I did the conversions to check how the dates operate, and wrote the code based on that

Comment (pandusonu2) [reply components/calendar/src/julian.rs:36]: Thanks for the above link, https://github.com/EdReingold/calendar-code2/blob/1ee51ecfaae6f856b0d7de3e36e9042100b4f424/calendar.l#L1704

I think that explains the edge case handling I had done in the `date_to_iso()`

Comment (Manishearth) [reply components/calendar/src/julian.rs:36]: Can you link to that source and write your math in a way that matches theirs, with explainer comments?

In particular I'm still not convinced that `calculate_day_difference()` calculates the slack for an _iso_ date, can always work when applied to a _julian_ date since it's going to have problems in the period where that slack shifts

**Comment by pandusonu2** [comment]
Ive added algorithms according to the book in 10585a5

But I don't really like it...... 
Alternative suggestion: revert 10585a5 and merge the code before that, and work on `julian<->fixed integer time<->iso` in another PR if we want to use that kind of conventions

File: components/calendar/src/iso.rs [hunk 14]
```diff
@@ -368,6 +368,68 @@ impl Iso {
         }
     }
 
+    pub(crate) fn fixed_from_iso(date: IsoDateInner) -> i32 {
```
Comment (Manishearth) [inline components/calendar/src/iso.rs:371]: nit: Please document this with something saying that "Fixed dates are represented by counting days since January 1, 1, with January 1, 1 being Fixed Date 1"

**Review by Manishearth — CHANGES_REQUESTED** [review]

File: components/calendar/src/julian.rs [hunk 15]
```diff
@@ -0,0 +1,380 @@
...
+        match u8::from(month) {
+            4 | 6 | 9 | 11 => 30,
+            2 if Self::is_leap_year(year) => 29,
+            2 => 28,
+            _ => 31,
+        }
+    }
+
+    fn fixed_from_julian(date: IsoDateInner) -> i32 {
+        let mut fixed: i32 = -1 + 365 * (date.year.0 - 1) + (date.year.0 - 1) / 4;
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply components/calendar/src/julian.rs:142]: Also I think this makes the first term -2? `julian-epoch` is `-1` i think

File: components/calendar/src/julian.rs [hunk 16]
```diff
@@ -0,0 +1,380 @@
...
+    fn fixed_from_julian_integers(year: i32, month: i32, day: i32) -> i32 {
+        Self::fixed_from_julian(
+            *Date::new_iso_date_from_integers(year, month as u8, day as u8)
+                .unwrap()
+                .inner(),
+        )
+    }
+
+    fn julian_from_fixed(date: i32) -> JulianDateInner {
+        let approx = 1464 + (4 * date) / 1461;
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:165]: issue: this is supposed to be `4 * date - julian_epoch`, where `julian_epoch` is the date of Julian Jan 1, 1, which is `-1`

File: components/calendar/src/julian.rs [hunk 17]
```diff
@@ -0,0 +1,380 @@
...
+        let year = if approx <= 0 { approx - 1 } else { approx };
+        let prior_days = date - Self::fixed_from_julian_integers(year, 1, 1);
+        let correction = if date < Self::fixed_from_julian_integers(year, 3, 1) {
+            0
+        } else if Self::is_leap_year(IsoYear::from(year)) {
+            1
+        } else {
+            2
+        };
+        let month = 12 * (prior_days + correction) / 367 + 373;
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:175]: the division by 367 is on the outside

File: components/calendar/src/iso.rs [hunk 18]
```diff
@@ -368,6 +368,68 @@ impl Iso {
...
+        let d1 = date % 146097;
+        let n_100 = d1 / 36524;
+        let d2 = d1 % 36524;
+        let n4 = d2 / 1461;
+        let d3 = d2 % 1461;
+        let n1 = d3 / 365;
+
+        let year = 400 * n_400 + 100 * n_100 + 4 * n4 + n1;
+
+        year
```
Comment (Manishearth) [inline components/calendar/src/iso.rs:411]: missing the last step for handling leap years: the result is `year + 1` when `n100 == 4 || n1 == 4`. This is clear in the lisp code but less clear in the book because the book writes this function out of order with the result up top

File: components/calendar/src/iso.rs [hunk 19]
```diff
@@ -368,6 +368,68 @@ impl Iso {
...
+        let year = Self::iso_year_from_fixed(date);
+        let prior_days = date - Self::iso_new_year(year);
+        let correction = if date < Self::fixed_from_iso_integers(year, 3, 1) {
+            0
+        } else if Self::is_leap_year(IsoYear::from(year)) {
+            1
+        } else {
+            2
+        };
+        let month = (12 * (prior_days + correction)) / 367 + 373;
```
Comment (Manishearth) [inline components/calendar/src/iso.rs:428]: the division by 367 should go outside

**Review by Manishearth — CHANGES_REQUESTED** [review]
Bunch of mistakes when copying over the algorithm, but the structure seems right!

**Comment by pandusonu2** [comment]
> Ive added algorithms according to the book in [10585a5](https://github.com/unicode-org/icu4x/commit/10585a5b7ccf0b7b1a24ed15c9be251f6c94c91c)
> 
> But I don't really like it...... Alternative suggestion: revert [10585a5](https://github.com/unicode-org/icu4x/commit/10585a5b7ccf0b7b1a24ed15c9be251f6c94c91c) and merge the code before that, and work on `julian<->fixed integer time<->iso` in another PR if we want to use that kind of conventions

I guess it was just a matter of rereading the algorithm multiple times till it made sense. Fine with the current changes (-.-)>

File: components/calendar/src/julian.rs [hunk 20]
```diff
@@ -8,6 +8,9 @@ use crate::iso::{Iso, IsoDateInner, IsoDay, IsoMonth, IsoYear};
 use crate::{types, Calendar, Date, DateDuration, DateDurationUnit, DateTime, DateTimeError};
 use core::convert::TryInto;
 
+// Julian epoch is equivalent to fixed_from_iso of December 30th of 0 year
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:11]: probably mention something like "January 1, 1 Julian = December 30, 1 BC Gregorian"

File: components/calendar/src/julian.rs [hunk 21]
```diff
@@ -138,16 +141,25 @@ impl Julian {
         }
     }
 
+    // Fixed is day count representation of calendars starting from Jan 1st of year 1 of Georgian Calendar.
+    // The fixed calculations algorithms are from the Calendrical Calculations book
+    //
+    // Lips code reference: https://github.com/EdReingold/calendar-code2/blob/1ee51ecfaae6f856b0d7de3e36e9042100b4f424/calendar.l#L1689-L1738
```
Comment (Manishearth) [inline components/calendar/src/julian.rs:147]: please document the other functions too

**Review by Manishearth — APPROVED** [review]
Your testcases are passing now, yes?

**Comment by pandusonu2** [comment]
> Your testcases are passing now, yes?

Yes

**Comment by Manishearth** [comment]
Alright, thanks! Hopefully once we get the CalendarArithmetic stuff done it becomes way easier to add calendars
